tianshou/data/buffer.py

```diff
@@ -464,69 +397,295 @@ def add(
+        episode_rewards = []  # (len(buffer_ids), ...)
+        for batch_idx, env_id in enumerate(buffer_ids):
+            length, reward = self.buffers[env_id].add(**batch[batch_idx])
+            episode_lengths.append(length)
+            episode_rewards.append(reward)
+        return np.stack(episode_lengths), np.stack(episode_rewards)
+
+    def sample_index(self, batch_size: int) -> np.ndarray:
+        if batch_size < 0:
+            return np.array([], np.int)
+        if self._sample_avail and self.stack_num > 1:
+            all_indices = np.concatenate([
+                buf.sample_index(0) + offset
+                for offset, buf in zip(self._offset, self.buffers)])
+            if batch_size == 0:
+                return all_indices
+            else:
+                return np.random.choice(all_indices, batch_size)
+        if batch_size == 0:  # get all available indices
+            sample_num = np.zeros(self.buffer_num, np.int)
+        else:
+            buffer_lens = np.array([len(buf) for buf in self.buffers])
+            buffer_idx = np.random.choice(self.buffer_num, batch_size,
+                                          p=buffer_lens / buffer_lens.sum())
+            sample_num = np.bincount(buffer_idx, minlength=self.buffer_num)
+            # avoid batch_size > 0 and sample_num == 0 -> get child's all data
+            sample_num[sample_num == 0] = -1
+
+        return np.concatenate([
+            buf.sample_index(bsz) + offset
+            for offset, buf, bsz in zip(self._offset, self.buffers, sample_num)
+        ])
+
+
+class CachedReplayBuffer(ReplayBuffers):
+    """CachedReplayBuffer contains a given main buffer and n cached buffers, \
+    cached_buffer_num * ReplayBuffer(size=max_episode_length).
+
+    The memory layout is: ``| main_buffer | cached_buffer[0] | cached_buffer[1]
+    | ... | cached_buffer[cached_buffer_num - 1]``.
+
+    The data is first stored in cached buffers. When the episode is
+    terminated, the data will move to the main buffer and the corresponding
+    cached buffer will be reset.
+
+    :param ReplayBuffer main_buffer: the main buffer whose ``.update()``
+        function behaves normally.
+    :param int cached_buffer_num: number of ReplayBuffer needs to be created
+        for cached buffer.
+    :param int max_episode_length: the maximum length of one episode, used in
+        each cached buffer's maxsize.
+
+    .. seealso::
+
+        Please refer to :class:`~tianshou.data.ReplayBuffer` or
+        :class:`~tianshou.data.ReplayBuffers` for more detailed
+        explanation.
+    """
+
+    def __init__(
+        self,
+        main_buffer: ReplayBuffer,
+        cached_buffer_num: int,
+        max_episode_length: int,
+    ) -> None:
+        assert cached_buffer_num > 0 and max_episode_length > 0
+        self.main_buffer = main_buffer
+        self._is_prioritized = isinstance(main_buffer, PrioritizedReplayBuffer)
+        kwargs = self.main_buffer.options
```

Comment:
I think self.cached_buffer_num = cached_buffer_num is needed because len(self.buffers) - 1 seems not good and unclear.
Also, suggest add self.cached_buffers = self.buffers[1:] (will be needed in later commit)